File: src/test/java/com/github/jactor/web/controller/HomeControllerTest.java
package com.github.jactor.web.controller;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.springframework.test.web.servlet.setup.MockMvcBuilders.standaloneSetup;

import com.github.jactor.web.i18n.MyMessages;
import com.github.jactor.web.model.HomePageModel;
import java.util.Map;
import java.util.Objects;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.InternalResourceViewResolver;

@SpringBootTest
@DisplayName("The HomeController")
class HomeControllerTest {

  private MockMvc mockMvc;

  @Autowired
  private MyMessages myMessages;
  @Value("${spring.mvc.view.prefix}")
  private String prefix;
  @Value("${spring.mvc.view.suffix}")
  private String suffix;

  @BeforeEach
  void mockMvcWithViewResolver() {
    InternalResourceViewResolver internalResourceViewResolver = new InternalResourceViewResolver();
    internalResourceViewResolver.setPrefix(prefix);
    internalResourceViewResolver.setSuffix(suffix);

    mockMvc = standaloneSetup(new HomeController(myMessages))
        .setViewResolvers(internalResourceViewResolver)
        .build();
  }

  @DisplayName("should create a homepage dto with my messages")
  @Test
  void shouldCreateHomepageDtoWithMyMessages() throws Exception {
    ModelAndView modelAndView = mockMvc.perform(
        get("/home")).andExpect(status().isOk()
    ).andReturn().getModelAndView();

    assertAll(
        () -> assertThat(modelAndView).as("modelAndView").isNotNull(),
        () -> {
          Map<String, Object> model = Objects.requireNonNull(modelAndView).getModel();

          assertThat(model).isNotNull();

          HomePageModel homePageModel = (HomePageModel) model.get("homepage");

          assertAll(
              () -> assertThat(homePageModel.getParagraphs()).as("paragraphs").hasSize(3),
              () -> assertThat(homePageModel.getTechnologies()).as("technologies").hasSize(8)
          );
        }
    );
  }
}

File: src/test/kotlin/com/github/jactor/web/controller/HomeControllerTest.kt
package com.github.jactor.web.controller

import com.github.jactor.web.i18n.MyMessages
import com.github.jactor.web.model.HomePageModel
import junit.framework.AssertionFailedError
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.servlet.view.InternalResourceViewResolver

@SpringBootTest
@DisplayName("The HomeController")
internal class HomeControllerTest {
    private lateinit var mockMvc: MockMvc

    @Autowired
    private lateinit var myMessages: MyMessages

    @Value("\${spring.mvc.view.prefix}")
    private lateinit var prefix: String

    @Value("\${spring.mvc.view.suffix}")
    private lateinit var suffix: String

    @BeforeEach
    fun `mock mvc with view resolver`() {
        val internalResourceViewResolver = InternalResourceViewResolver()

        internalResourceViewResolver.setPrefix(prefix)
        internalResourceViewResolver.setSuffix(suffix)

        mockMvc = MockMvcBuilders.standaloneSetup(HomeController(myMessages))
            .setViewResolvers(internalResourceViewResolver)
            .build()
    }

    @Test
    fun `should create a homepage dto with my messages`() {
        val modelAndView = mockMvc.perform(
            MockMvcRequestBuilders.get("/home")
        ).andExpect(
            MockMvcResultMatchers.status().isOk
        ).andReturn().modelAndView

        assertAll(
            { assertThat(modelAndView).`as`("modelAndView").isNotNull() },
            {
                val model = modelAndView?.model ?: throw AssertionFailedError("No model to be found!")
                assertThat(model).isNotNull

                val homePageModel = model["homepage"] as HomePageModel?
                assertThat(homePageModel!!.technologies).`as`("technologies").hasSize(10)
            }
        )
    }
}